src/crypto/eddsa.rs

```diff
@@ -48,62 +58,57 @@ impl Signer for SecretKey {
 }
 
 /// Generate an EdDSA KeyPair.
-pub fn generate_key<R: Rng + CryptoRng>(mut rng: R) -> (PublicParams, PlainSecretParams) {
+pub fn generate_key<R: Rng + CryptoRng>(
+    mut rng: R,
+    mode: Mode,
+) -> (PublicParams, PlainSecretParams) {
     let mut bytes = Zeroizing::new([0u8; ed25519_dalek::SECRET_KEY_LENGTH]);
     rng.fill_bytes(&mut *bytes);
     let secret = ed25519_dalek::SigningKey::from_bytes(&bytes);
     let public = ed25519_dalek::VerifyingKey::from(&secret);
 
-    // public key
-    let mut q = Vec::with_capacity(33);
-    q.push(0x40);
-    q.extend_from_slice(&public.to_bytes());
-
-    // secret key
-    let p = Mpi::from_raw_slice(&secret.to_bytes());
-    bytes.zeroize();
-
-    (
-        PublicParams::EdDSALegacy {
-            curve: ECCCurve::Ed25519,
-            q: q.into(),
-        },
-        PlainSecretParams::EdDSALegacy(p),
-    )
+    match mode {
+        Mode::EdDSALegacy => {
+            // public key
+            let mut q = Vec::with_capacity(33);
+            q.push(0x40);
+            q.extend_from_slice(&public.to_bytes());
+
+            // secret key
+            let p = Mpi::from_raw_slice(&secret.to_bytes());
+            bytes.zeroize();
+
+            (
+                PublicParams::EdDSALegacy {
+                    curve: ECCCurve::Ed25519,
+                    q: q.into(),
+                },
+                PlainSecretParams::EdDSALegacy(p),
+            )
+        }
+        Mode::Ed25519 => (
+            PublicParams::Ed25519 {
+                public: public.to_bytes(),
+            },
+            PlainSecretParams::Ed25519(secret.to_bytes()),
+        ),
+    }
 }
 
 /// Verify an EdDSA signature.
 pub fn verify(
     curve: &ECCCurve,
-    q: &[u8],
+    public: &[u8],
     _hash: HashAlgorithm,
     hashed: &[u8],
-    sig: &[Mpi],
+    sig_bytes: &[u8],
 ) -> Result<()> {
     match *curve {
         ECCCurve::Ed25519 => {
-            ensure_eq!(sig.len(), 2);
-
-            let r = sig[0].as_bytes();
-            let s = sig[1].as_bytes();
-
-            ensure!(r.len() < 33, "invalid R (len)");
-            ensure!(s.len() < 33, "invalid S (len)");
-            ensure_eq!(q.len(), 33, "invalid Q (len)");
-            ensure_eq!(q[0], 0x40, "invalid Q (prefix)");
```

Comment:
where did this logic go?

Reply:
To `src/composed/signed_key/public.rs`. My idea was that the "legacy" format specific stuff is abstracted away there, while `eddsa.rs` can more format-agnostic.

(Not sure if that's the right idea though, seeing how key generation in essda.rs does OpenPGP format-specific things :thinking:)

Reply:
ah okay, makes sense, just trying to understand